linera-client/src/chain_listener.rs

```diff
@@ -274,37 +199,143 @@ impl<C: ClientContext> ChainClientListener<C> {
+                    let Some(notification) = maybe_notification else {
+                        let chain_id = *self.listening.keys().nth(index).unwrap();
+                        self.listening.remove(&chain_id);
+                        warn!("Notification stream for {chain_id} closed");
+                        continue;
+                    };
+                    return Ok(Action::Notification(notification));
+                }
+            }
+        }
+    }
+
+    /// Returns the next timeout to process, and the chain to which it applies.
+    fn next_timeout(&self) -> Result<(ChainId, Timestamp), Error> {
+        let (chain_id, client) = self
+            .listening
+            .iter()
+            .min_by_key(|(_, client)| client.timeout)
+            .expect("No chains left to listen to");
+        Ok((*chain_id, client.timeout))
+    }
+
+    /// Updates the validators about the chain.
+    async fn update_validators(&self, notification: &Notification) -> Result<(), Error> {
+        let chain_id = notification.chain_id;
+        let listening_client = self.listening.get(&chain_id).expect("missing client");
+        if let Err(error) = listening_client.client.update_validators(None).await {
+            warn!(
+                "Failed to update validators about the local chain after \
+                 receiving {notification:?} with error: {error:?}"
+            );
+        }
+        Ok(())
+    }
+
+    /// Updates the wallet based on the client for this chain.
+    async fn update_wallet(&self, chain_id: ChainId) -> Result<(), Error> {
+        let client = &self
+            .listening
+            .get(&chain_id)
+            .expect("missing client")
+            .client;
+        self.context.lock().await.update_wallet(client).await?;
+        Ok(())
+    }
+
     /// Processes the inbox, unless `skip_process_inbox` is set.
     ///
     /// If no block can be produced because we are not the round leader, a timeout is returned
     /// for when to retry; otherwise `u64::MAX` is returned.
     ///
     /// The wallet is persisted with any blocks that processing the inbox added. An error
     /// is returned if persisting the wallet fails.
-    async fn maybe_process_inbox(&mut self) -> Result<(), Error> {
-        self.timeout = Timestamp::from(u64::MAX);
-        if self.listener.config.skip_process_inbox {
+    async fn maybe_process_inbox(&mut self, chain_id: ChainId) -> Result<(), Error> {
+        if self.config.skip_process_inbox {
             debug!("Not processing inbox due to listener configuration");
             return Ok(());
         }
+        let listening_client = self.listening.get_mut(&chain_id).expect("missing client");
         debug!("Processing inbox");
-        match self.client.process_inbox_without_prepare().await {
+        listening_client.timeout = Timestamp::from(u64::MAX);
+        match listening_client
+            .client
+            .process_inbox_without_prepare()
```

Comment:
Does it mean that we can finally make client functions like `process_inbox_without_prepare ` take `&mut self` again?

Reply:
I'm not sure. Was the chain listener the only reason we changed that, @Twey?

Reply:
Pedantically we can, but it wouldn't change much semantically since chain clients can be taken from the `Client` with a mere `&` reference.  This is probably still necessary because `ChainClient::listen` still shares the client.